pkg/dyninst/decode/marshal.go

```diff
@@ -0,0 +1,473 @@
+		}
+		address := binary.NativeEndian.Uint64(data[0:8])
+		length := binary.NativeEndian.Uint64(data[8:16])
+		if length == 0 {
+			if err := writeTokens(enc,
+				jsontext.BeginArray,
+				jsontext.EndArray); err != nil {
+				return err
+			}
+			return nil
+		}
+		elementType := v.Data.ID
+		elementSize := int(v.Data.Element.GetByteSize())
+		sliceDataItem, ok := dataItems[typeAndAddr{
+			addr:   address,
+			irType: uint32(elementType),
+		}]
+		if !ok {
+			return errors.New("slice data item not found in data items")
+		}
+		var err error
+		if err = writeTokens(enc, jsontext.BeginArray); err != nil {
+			return err
+		}
+		sliceLength := int(sliceDataItem.Header().Length) / elementSize
+		sliceData := sliceDataItem.Data()
+		for i := range int(sliceLength) {
+			elementData := sliceData[i*elementSize : (i+1)*elementSize]
+			if err := d.encodeValue(enc,
+				dataItems, currentlyEncoding, v.Data.Element, elementData); err != nil {
+				return err
+			}
+		}
+		if err = writeTokens(enc, jsontext.EndArray); err != nil {
+			return err
+		}
+		return nil
+	case *ir.GoChannelType:
+		return writeTokens(enc, jsontext.String("unimplemented - channel type"))
+	case *ir.GoStringHeaderType:
+		var address uint64
+		for _, field := range v.Fields {
+			if field.Name == "str" {
+				if uint32(field.Type.GetByteSize()) != 8 || len(data) < int(field.Offset+field.Type.GetByteSize()) {
+					return errors.New("malformed string field 'str'")
+				}
+				address = binary.NativeEndian.Uint64(data[field.Offset : field.Offset+uint32(field.Type.GetByteSize())])
+			}
+		}
+		stringValue, ok := dataItems[typeAndAddr{
+			irType: uint32(v.Data.GetID()),
+			addr:   address,
+		}]
+		if !ok {
+			return fmt.Errorf("string content not present in data items")
+		}
+		length := stringValue.Header().Length
+		if len(stringValue.Data()) < int(length) {
+			return errors.New("string content not long enough for known length")
+		}
+		if err := writeTokens(enc,
+			jsontext.String(string(stringValue.Data()[:length]))); err != nil {
+			return err
+		}
+		return nil
+	case *ir.GoMapType:
+		return writeTokens(enc, jsontext.String("unimplemented - map type"))
+	case *ir.GoHMapHeaderType:
+		return writeTokens(enc, jsontext.String("unimplemented - hmap header type"))
+	case *ir.GoHMapBucketType:
+		return writeTokens(enc, jsontext.String("unimplemented - hmap bucket type"))
+	case *ir.GoSwissMapHeaderType:
+		return writeTokens(enc, jsontext.String("unimplemented - swiss map header type"))
+	case *ir.GoSwissMapGroupsType:
+		return writeTokens(enc, jsontext.String("unimplemented - swiss map groups type"))
+	case *ir.EventRootType:
+		return writeTokens(enc, jsontext.String("unimplemented - event root type"))
+	default:
+		return errors.New("invalid type")
+	}
+	return nil
+}
+
+func (d *Decoder) encodeBaseTypeValue(enc *jsontext.Encoder, irType *ir.BaseType, data []byte) error {
+	kind, ok := irType.GetGoKind()
+	if !ok {
+		return errors.New("No go kind")
```

Comment:
Give us more information about the type and its name. Also, nit, error messages shouldn't start with capitals so they compose better